concordium-consensus/src/Concordium/GlobalState/Persistent/LMDB.hs

```diff
@@ -80,30 +83,56 @@ data StoredBlock pv st = StoredBlock {
   sbState :: !st
 }
 
--- Note: 'openReadOnlyDatabase' works on the presumption that the state is always the last part of
--- the serialization, so we can serialize a stored block with any state type and deserialize it
--- with the unit state type.  Any changes to the serialization used here must respect this or
--- be accompanied by corresponding changes there.
-instance (IsProtocolVersion pv, S.Serialize st) => S.Serialize (StoredBlock pv st) where
-  put StoredBlock{..} = S.put sbFinalizationIndex <>
-          S.put sbInfo <>
-          putBlock (protocolVersion @pv) sbBlock <>
-          S.put sbState
-  get = do
-          sbFinalizationIndex <- S.get
-          sbInfo <- S.get
-          sbBlock <- getBlock (protocolVersion @pv) (utcTimeToTransactionTime (_bpReceiveTime sbInfo))
-          sbState <- S.get
-          return StoredBlock{..}
+putStoredBlock :: forall pv st . (IsProtocolVersion pv, S.Serialize st) => S.Putter (StoredBlock pv st)
+putStoredBlock StoredBlock{..} =
+      S.put sbFinalizationIndex <>
+      S.put sbInfo <>
+      putBlock (protocolVersion @pv) sbBlock <>
+      S.put sbState
 
 -- |A block store table. A @BlockStore pv st@ stores @StoredBlock pv st@ blocks
 -- indexed by 'BlockHash'.
 newtype BlockStore (pv :: ProtocolVersion) st = BlockStore MDB_dbi'
 
-instance (IsProtocolVersion pv, S.Serialize st) => MDBDatabase (BlockStore pv st) where
+-- |A refinement of 'Serialize' that indicates that the size (i.e., amount of
+-- bytes used) of the serialized value is independend of the value. This must be
+-- exact, not just an upper bound.
+class S.Serialize a => FixedSizeSerialization a where
+  -- |Return the size of serialized values in bytes.
+  serializedSize :: Proxy a -> Int
+
+instance FixedSizeSerialization () where
+  serializedSize _ = 0
+
+instance FixedSizeSerialization (BlobRef a) where
+  serializedSize _ = 8
+
+-- This instance is needed for paired state.
+instance (FixedSizeSerialization a, FixedSizeSerialization b) => FixedSizeSerialization (a, b) where
+  serializedSize _ = serializedSize (Proxy @a) + serializedSize (Proxy @b)
+
+instance (IsProtocolVersion pv, FixedSizeSerialization st) => MDBDatabase (BlockStore pv st) where
   type DBKey (BlockStore pv st) = BlockHash
   type DBValue (BlockStore pv st) = StoredBlock pv st
   encodeKey _ = hashToByteString . blockHash
+  encodeValue _ sb = S.runPutLazy (putStoredBlock sb)
+  decodeValue _ blockHash mdbVal = do
+    bs <- unsafeByteStringFromMDB_val mdbVal
```

Comment:
Even if the resulting type is harboring bytestrings it is generally not an issue since `getByteString` creates a copy. We would have to be using `getBytes` directly and retain the resulting bytestring to retain pointers into the source, but I don't see us doing that.

While I see that it is a bit risky, I think the particular use that we have is OK and avoids copying potentially a large amount of data for genesis blocks.

I have provided more documentation.